File: examples/typescript/react/notes-showcase/src/models/noteRepository.ts
/**
 * INoteRepository — persistence port for notebooks and notes.
 *
 * Mirrors the C# / Python `INoteRepository` contract. All methods are async
 * and return Promises so views never block. Simulated I/O delays in
 * `InMemoryNoteRepository` mirror the canonical 300/150/200/120/120/150 ms
 * timings from scenario doc §7.
 */
import type { NotebookModel } from "./notebookModel.js";
import type { NoteModel } from "./noteModel.js";

export interface RepositorySnapshot {
  readonly notebooks: readonly NotebookModel[];
  readonly notes: readonly NoteModel[];
}

export interface INoteRepository {
  /** Loads everything. ~300 ms simulated delay. */
  loadAll(): Promise<RepositorySnapshot>;
  /** Loads notes belonging to a given notebook id. ~150 ms simulated delay. */
  loadNotes(notebookId: string): Promise<NoteModel[]>;
  /** Inserts or updates a note (stamps updatedAt). ~200 ms simulated delay. */
  saveNote(note: NoteModel): Promise<void>;
  /** Removes the note with the given id (no-op if absent). ~120 ms simulated delay. */
  deleteNote(id: string): Promise<void>;
  /** Appends a new notebook to the store. ~120 ms simulated delay. */
  addNotebook(notebook: NotebookModel): Promise<void>;
  /** Writes a JSON snapshot to the given path. ~150 ms simulated delay. */
  export(
    notebooks: readonly NotebookModel[],
    notes: readonly NoteModel[],
    path: string,
  ): Promise<void>;
}

File: examples/python/textual/notes_showcase/src/notes_showcase/models/note_repository.py
"""INoteRepository — async persistence port for the showcase.

Mirrors the C# ``INoteRepository`` contract (scenario §6.2). The protocol is
runtime-checkable so test doubles satisfy it without explicit inheritance.
"""

from __future__ import annotations

from typing import Protocol, runtime_checkable

from notes_showcase.models.note_model import NoteModel
from notes_showcase.models.notebook_model import NotebookModel


@runtime_checkable
class INoteRepository(Protocol):
    """Async persistence port for notebooks and notes."""

    async def load_all(self) -> tuple[list[NotebookModel], list[NoteModel]]:
        """Return all notebooks + all notes (simulated ~300 ms latency)."""
        ...

    async def load_notes(self, notebook_id: str) -> list[NoteModel]:
        """Return notes belonging to *notebook_id* (simulated ~150 ms latency)."""
        ...

    async def search_notes(
        self, term: str, token: str | None, page_size: int
    ) -> tuple[list[NoteModel], str | None]:
        """Search all notes with opaque forward-only token paging."""
        ...

    async def save_note(self, note: NoteModel) -> None:
        """Insert or update *note* (simulated ~200 ms latency)."""
        ...

    async def delete_note(self, note_id: str) -> None:
        """Remove the note with id *note_id* (simulated ~120 ms latency)."""
        ...

    async def add_notebook(self, notebook: NotebookModel) -> None:
        """Append *notebook* to the store (simulated ~120 ms latency)."""
        ...

    async def export(
        self,
        notebooks: list[NotebookModel],
        notes: list[NoteModel],
        path: str,
    ) -> None:
        """Write a JSON snapshot to *path* (simulated ~150 ms latency)."""
        ...
